File: Scene.cpp
//
// Created by Göksu Güvendiren on 2019-05-14.
//

#include "Scene.hpp"
#include "Material.hpp"


void Scene::buildBVH() {
    printf(" - Generating BVH...\n\n");
    this->bvh = new BVHAccel(objects, 1, BVHAccel::SplitMethod::NAIVE);
}

Intersection Scene::intersect(const Ray &ray) const
{
    return this->bvh->Intersect(ray);
}

void Scene::sampleLight(Intersection &pos, float &pdf) const
{
    float emit_area_sum = 0;
    for (uint32_t k = 0; k < objects.size(); ++k) {
        if (objects[k]->hasEmit()){
            emit_area_sum += objects[k]->getArea();
        }
    }
    float p = get_random_float() * emit_area_sum;
    emit_area_sum = 0;
    for (uint32_t k = 0; k < objects.size(); ++k) {
        if (objects[k]->hasEmit()){
            emit_area_sum += objects[k]->getArea();
            if (p <= emit_area_sum){
                objects[k]->Sample(pos, pdf);
                break;
            }
        }
    }
}

bool Scene::trace(
        const Ray &ray,
        const std::vector<Object*> &objects,
        float &tNear, uint32_t &index, Object **hitObject)
{
    *hitObject = nullptr;
    for (uint32_t k = 0; k < objects.size(); ++k) {
        float tNearK = kInfinity;
        uint32_t indexK;
        Vector2f uvK;
        if (objects[k]->intersect(ray, tNearK, indexK) && tNearK < tNear) {
            *hitObject = objects[k];
            tNear = tNearK;
            index = indexK;
        }
    }


    return (*hitObject != nullptr);
}


// Implementation of Path Tracing
Vector3f Scene::castRay(const Ray &ray, int depth) const
{
    // TODO Implement Path Tracing Algorithm here
    // shade(p, wo)
    //  sampleLight(inter , pdf_light)
    //  Get x, ws, NN, emit from inter
    //  Shoot a ray from p to x
    //  If the ray is not blocked in the middle
    //  L_dir = emit * eval(wo, ws, N) * dot(ws, N) * dot(ws, NN) / |x-p|^2 / pdf_light
    Vector3f L_dir = 0.f;

    // initial hit point
    Intersection p = intersect(ray);

    if (!p.happened)
    {
        return Vector3f(0.f, 0.f, 0.f);
    }

    // If we directly shoot the light area, we return the emit energy.
    // This is L_e term in the full rendering equation.
    if (p.m->hasEmission())
    {
        if (depth == 0)
        {
            return p.m->getEmission();
        }
        else
        {
            return Vector3f(0.f, 0.f, 0.f);
        }
    }

    // outgoing direction
    Vector3f wo = (ray.origin - p.coords).norm();

    /*
     * Direct Lighting
     * We do light sampling here.
     */

    // x is sample point on scene area light
    Intersection x;
    float pdf_arealight = 0.f;
    sampleLight(x, pdf_arealight);

    // Direction between hit point and light samples.
    Vector3f ws = normalize(x.coords - p.coords);

    // We trace shadow ray to determine the direct lighting.
    Ray shadowRay(p.coords + ws * 0.0005f, ws);
    Intersection isBlock = intersect(shadowRay);

    // Check if shadow ray's hit point is light sample x
    // If there's occlusion cause by other object, the p is in shadow (related to x), so L_dir = 0.
    if (isBlock.happened && (isBlock.coords - x.coords).norm() < 0.001f)
    {
        L_dir = x.emit
                * p.m->eval(wo, ws, p.normal)
                * dotProduct(ws, p.normal)
                * dotProduct(-ws, x.normal)
                / ((x.coords - p.coords).norm() * (x.coords - p.coords).norm())
                / pdf_arealight;
    }
    else
    {
        L_dir = 0.f;
    }

    /*
     * Indirect Lighting
     * multi-bounce indirect illumination
     */

    //  L_indir = 0.0
    //  Test Russian Roulette with probability RussianRoulette
    //  wi = sample(wo, N)
    //  Trace a ray r(p, wi)
    //  If ray r hit a non -emitting object at q
    //  L_indir = shade(q, wi) * eval(wo, wi, N) * dot(wi, N) / pdf(wo, wi, N) / RussianRoulette
    //  Return L_dir + L_indir
    Vector3f L_indir = 0.f;

    // min bounces: 4
    if (depth > 4)
    {
        if (get_random_float() > RussianRoulette)
        {
            return L_dir;
        }
    }

    Vector3f wi = p.m->sample(wo, p.normal);
    float cos_theta = dotProduct(wi, p.normal);
    float pdf_indir = cos_theta / M_PI;

    // Ensure that cos_theta > 0 (since we are in a Hemispherical Integral Domain) and pdf > 0.
    if (pdf_indir > 0.f && cos_theta > 0.f)
    {
        Vector3f shading = castRay(Ray(p.coords + wi * 0.0005f, wi), depth + 1);
        // Intersection q = intersect(Ray(p.coords, wi));

        // Indirect illumination should not add emit itself directly.
        // We trace 4-bounce ray for every pixel, when it comes to >4, we use RR to terminate.
        float rr = (depth > 4) ? RussianRoulette : 1.f;

        L_indir = shading * p.m->eval(wo, wi, p.normal) * cos_theta
                / pdf_indir
                / rr;
    }

    return L_dir + L_indir;
}

File: rust/src/scene.rs
use std::sync::Arc;

use crate::bvh::BVHAccel;
use crate::global::{get_random_float, M_PI};
use crate::intersection::Intersection;
use crate::object::Hittable;
use crate::ray::Ray;
use crate::vector::Vec3;

pub struct Scene {
    pub width: usize,
    pub height: usize,
    pub fov: f32,
    pub background_color: Vec3,
    pub max_depth: i32,
    pub russian_roulette: f32,
    pub objects: Vec<Arc<dyn Hittable>>,
    pub bvh: Option<BVHAccel>,
}

impl Scene {
    pub fn new(width: usize, height: usize) -> Self {
        Self {
            width,
            height,
            fov: 40.0,
            background_color: Vec3::new(0.235294, 0.67451, 0.843137),
            max_depth: 1,
            russian_roulette: 0.8,
            objects: Vec::new(),
            bvh: None,
        }
    }

    pub fn add(&mut self, object: Arc<dyn Hittable>) {
        self.objects.push(object);
    }

    pub fn build_bvh(&mut self) {
        println!(" - Generating BVH...\n");
        self.bvh = Some(BVHAccel::new(self.objects.clone()));
    }

    pub fn intersect(&self, ray: &Ray) -> Intersection {
        if let Some(bvh) = &self.bvh {
            bvh.intersect(ray)
        } else {
            Intersection::default()
        }
    }

    pub fn sample_light(&self, pos: &mut Intersection, pdf: &mut f32) {
        let mut emit_area_sum = 0.0;
        for object in &self.objects {
            if object.has_emit() {
                emit_area_sum += object.get_area();
            }
        }

        let p = get_random_float() * emit_area_sum;
        emit_area_sum = 0.0;

        for object in &self.objects {
            if object.has_emit() {
                emit_area_sum += object.get_area();
                if p <= emit_area_sum {
                    object.sample(pos, pdf);
                    break;
                }
            }
        }
    }

    pub fn cast_ray(&self, ray: &Ray, depth: i32) -> Vec3 {
        let p = self.intersect(ray);
        if !p.happened {
            return Vec3::zero();
        }

        let material = p.material.as_ref().expect("material missing at hit point");

        if material.has_emission() {
            if depth == 0 {
                return material.get_emission();
            }
            return Vec3::zero();
        }

        let wo = (ray.origin - p.coords).normalize();

        let mut x = Intersection::default();
        let mut pdf_area_light = 0.0f32;
        self.sample_light(&mut x, &mut pdf_area_light);

        let ws = (x.coords - p.coords).normalize();
        let shadow_ray = Ray::new(p.coords + ws * 0.0005, ws);
        let is_block = self.intersect(&shadow_ray);

        let l_dir = if is_block.happened && (is_block.coords - x.coords).norm() < 0.001 {
            x.emit * material.eval(wo, ws, p.normal) * ws.dot(p.normal) * (-ws).dot(x.normal)
                / (x.coords - p.coords).norm().powi(2)
                / pdf_area_light
        } else {
            Vec3::zero()
        };

        if depth > 4 && get_random_float() > self.russian_roulette {
            return l_dir;
        }

        let wi = material.sample(wo, p.normal);
        let cos_theta = wi.dot(p.normal);
        let pdf_indir = cos_theta / M_PI;

        let l_indir = if pdf_indir > 0.0 && cos_theta > 0.0 {
            let shading = self.cast_ray(&Ray::new(p.coords + wi * 0.0005, wi), depth + 1);
            let rr = if depth > 4 {
                self.russian_roulette
            } else {
                1.0
            };
            shading * material.eval(wo, wi, p.normal) * cos_theta / pdf_indir / rr
        } else {
            Vec3::zero()
        };

        l_dir + l_indir
    }
}
